internal/engine/planner/lambda.go

```diff
@@ -4,54 +4,156 @@
-	iterRange  *exprpb.Expr
-	lambdaExpr *exprpb.Expr
-	operator   string
-	iterVar    string
+	iterRange *exprpb.Expr
+	expr      *exprpb.Expr
+	expr2     *exprpb.Expr
+	iterVar   string
+	iterVar2  string
+	operator  string
+}
+
+func getTransformMapExpression(w *wrapper) (string, *exprpb.Expr, error) {
+	const nTransformMapEntryArgs = 2
+	const nTransformMapArgs = 3
+	n := w.getArgsLen()
+	var op string
+
+	switch n {
+	case nTransformMapArgs:
+		op = TransformMap
+	case nTransformMapEntryArgs:
+		op = TransformMapEntry
+	default:
+		return "", nil, fmt.Errorf("unexpected args #, got: %d", n)
+	}
+	return op, w.getArg(n - 1).e(), nil
+}
+
+var ErrExpectedSortBy = errors.New("expected sortBy comprehension")
+
+const sortByFuncName = "sortBy"
+
+func mkSortByAST(e *exprpb.Expr_Comprehension) (*lambdaAST, error) {
+	const function = "@sortByAssociatedKeys"
+	w := (*wrapper)(e.Result)
+	var e2 *exprpb.Expr_Comprehension
+	if e2 = w.getArg(0).e().GetComprehensionExpr(); e2 == nil || w.getArgsLen() != 1 || w.Function() != function {
+		return nil, fmt.Errorf("%w, got %s", ErrExpectedSortBy, e.String())
+	}
+	obj := &lambdaAST{
+		operator:  sortByFuncName,
+		iterRange: e.AccuInit,
+		iterVar:   e2.IterVar,
+		expr:      (*wrapper)(e2.LoopStep).getArg(1).getListElement(0).e(),
+	}
+	if obj.expr == nil {
+		return nil, fmt.Errorf("%w, got %s", ErrExpectedSortBy, e.String())
+	}
+	return obj, nil
 }
 
 func buildLambdaAST(e *exprpb.Expr_Comprehension) (*lambdaAST, error) {
+	var function string
+	var loopStep *wrapper
+	switch ls := e.LoopStep.ExprKind.(type) {
+	case *exprpb.Expr_CallExpr:
+		function = ls.CallExpr.Function
+		loopStep = (*wrapper)(e.LoopStep)
+	case *exprpb.Expr_IdentExpr:
+		return mkSortByAST(e)
+	default:
+		return nil, fmt.Errorf("expected loop-step expression type CallExpr, got: %T", ls)
+	}
 	obj := &lambdaAST{
 		iterVar:   e.IterVar,
+		iterVar2:  e.IterVar2,
 		iterRange: e.IterRange,
 	}
-	var step *exprpb.Expr_CallExpr
-	var ok bool
-	if step, ok = e.LoopStep.ExprKind.(*exprpb.Expr_CallExpr); !ok {
-		return nil, fmt.Errorf("expected loop-step expression type CallExpr, got: %T", e.LoopStep.ExprKind)
-	}
-	switch step.CallExpr.Function {
+	switch function {
 	case operators.LogicalAnd:
 		obj.operator = All
-		obj.lambdaExpr = step.CallExpr.Args[1]
+		obj.expr = loopStep.getArg(1).e()
 	case operators.LogicalOr:
 		obj.operator = Exists
-		obj.lambdaExpr = step.CallExpr.Args[1]
+		obj.expr = loopStep.getArg(1).e()
 	case operators.Add:
 		obj.operator = Map
-		if elements := step.CallExpr.Args[1].GetListExpr().GetElements(); len(elements) > 0 {
-			obj.lambdaExpr = elements[0]
+		if obj.iterVar2 != "" {
+			obj.operator = TransformList
 		}
+		obj.expr = loopStep.getArg(1).getListElement(0).e()
 	case operators.Conditional:
 		switch e.AccuInit.ExprKind.(type) {
 		case *exprpb.Expr_ListExpr:
-			obj.operator = Filter
+			if e2 := loopStep.getArg(1).getArg(1).getListElement(0).e(); e2.GetCallExpr() != nil {
```

Comment:
Is it guaranteed that `getArg` never returns `nil` in this case? According to the implementation of `getArg`, it _could_  return `nil`...

There are a bunch of places in this file where method calls are chained without nil checks. Can we be absolutely sure that they won't panic at runtime?

Reply:
No, it will not panic in the presence of a `nil` pointer. I've added the test `TestWrapperDoNotPanic` to demonstrate this.